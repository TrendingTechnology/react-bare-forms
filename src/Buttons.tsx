import {ReactElement, default as React, Consumer} from "react";
import {FormConsumer, IFormContext} from "./form";

export interface IButton<T> {
    create(): (props: ISubmitButtonProps) => ReactElement<T>;
}


abstract class Button<T> implements IButton<T> {

    public abstract create(): (props: T) => ReactElement<T>;

}

export interface ISubmitButtonProps {
    text?: string;
    disabled?: false;
    className?: string;
    children?: any;
}

/**
 * SubmitButton is the main component for submitting forms.
 *
 */
class _SubmitButton implements Button<ISubmitButtonProps> {

    create() {
       return (props: ISubmitButtonProps) => {
           const { className = "btn btn-success btn-lg", disabled = true} = props;
           return (
               <FormConsumer>
                   {(context: IFormContext) => {
                       let isDisabled;
                       if(context && context.metadata && Object.keys(context.metadata.inputs).length > 0) {
                           const inputItem = context.metadata.inputs;
                           Object.keys(inputItem).map((key: any) => {
                               if(!inputItem[key].isValid) {
                                   isDisabled = true;
                               }
                           });
                       }
                       return (<button
                           disabled={isDisabled}
                           className={className}
                       >{props.children}</button>);
                   }}
               </FormConsumer>
           );
       }
    }
}

export const SubmitButton = new _SubmitButton().create();
