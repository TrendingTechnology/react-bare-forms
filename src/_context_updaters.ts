// Context Updaters
// Methods to update & track the state of all inputs, validators & general form metadata
//
// For example:
// const context = {
//     // other props...
//     metadata: {
//          inputs: {
//              email: {
//                  isValid: false,
//                  messages: ['Cannot be empty'],
//                  value: '',
//                  isTouched: false,
//              },
//              username: {
//                  isValid: true,
//                  messages: [],
//                  value: 'joebloggs',
//                  isTouched: true,
//              }
//          }
//     }
// };
import {IValidation} from "./validators";


export const updateValidationMetadata = (context: any, update: any) =>
    (fieldName: string, fieldValue: any, validation: IValidation): void => {
        if(typeof context.metadata === "undefined") {
            return;
        }
        if(!(fieldName in context.metadata)) {
            update({
                ...context,
                metadata: {
                    ...context.metadata,
                    [fieldName]: {
                        ...validation,
                        value: fieldValue,
                        isTouched: true,
                    }
                },
            });
            // Should update context.metadata value
        } else if(context.metadata[fieldName].value != fieldValue) {
            update({
                ...context,
                metadata: {
                    ...context.metadata,
                    [fieldName]: {
                        ...validation,
                        value: fieldValue,
                        isTouched: true,
                    }
                },
            });
        }

    };


