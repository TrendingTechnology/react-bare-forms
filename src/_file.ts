import {default as React} from "react";
import {getFileFromRef} from "./uncrontrolled";

// File Helper Functions


/** @internal **/
export interface IFile {
    readonly name: string;
    readonly lastModified: string;
    readonly lastModifiedDate: string;
    readonly size: string;
    readonly type: string;
}

/** @internal */
export function createFileObject(ref: React.RefObject<HTMLFormElement>): IFile {
    try {
        let {name, lastModified, lastModifiedDate, size, type} = getFileFromRef(ref);
        return {
            name,
            lastModified,
            lastModifiedDate,
            size,
            type,
        };
    } catch (err) {
          return null;
    }
}
