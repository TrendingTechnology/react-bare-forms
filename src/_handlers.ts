import * as React from "react";

/** @internal */
type TypeHandler = (e: React.ChangeEvent<any>, name: string) => void;

/** @internal */
export function updateStateFromPassedInContext(parentState: any, setParentState: Function): TypeHandler {
    return (e: React.ChangeEvent<any>, name: string) => {
        setParentState({
            [name]: e.target.value,
        });
    };
}
