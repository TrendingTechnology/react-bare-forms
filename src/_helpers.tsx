import * as React from "react";
import {handleChange} from "./handlers";
import {FieldTypes} from "./form-elements";
import {IFormContext} from "./form";
import {IValidation} from "./validators";

/**
 *
 * @param fieldType
 * @param name
 * @param context
 */
export const selectTextField = (fieldType: FieldTypes, name: any, context: IFormContext) => {
    switch(fieldType) {
        case "text": {
            return <input type="text" className="form-control" onChange={handleChange(name, context)} />;
        }
        case "email": {
            return <input type="email" className="form-control" onChange={handleChange(name, context)} />;
        }
        case "password": {
            // TODO confirm password
            return <input type="password" className="form-control" onChange={handleChange(name, context)} />;
        }
        default: {
            return <input type="text" className="form-control" onChange={handleChange(name, context)} />;
        }
    }
};

/**
 *
 * @param validationResult
 * @param context
 */
export const shouldShowValidation = (validationResult: IValidation, context: IFormContext) => {
    return (!validationResult.isValid && context.dynamic) ||
        (!validationResult.isValid && context.isSubmitted);
};


/**
 *
 * @param update
 * @param current
 */
export function setFormData(update: Function, current: IFormContext) {
    return (name: any, value: any) => {
        update({
            ...current,
            state: {
                ...current.state,
                [current.formKey]: {
                    ...current.state[current.formKey],
                    [name]: value,
                }
            }
        });
    };
}
