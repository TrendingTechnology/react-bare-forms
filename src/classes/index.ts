import {AbstractField} from "./_AbstractField";
import {AbstractMetadata} from "./_AbstractMetadata";
import {CheckBoxField} from "./_CheckBoxField";
import {FileField} from "./_FileField";
import {InputField} from "./_InputField";
import {Metadata} from "./_Metadata";
import {MetadataFile} from "./_MetadataFile";
import {MetadataGroup} from "./_MetadataGroup";
import {RadioField} from "./_RadioField";
import {SelectField} from "./_SelectField";
import {SubmitButton} from "./_SubmitButton";
import {TextAreaField} from "./_TextAreaField";


export {
    AbstractField,
    AbstractMetadata,
    CheckBoxField,
    FileField,
    InputField,
    Metadata,
    MetadataFile,
    MetadataGroup,
    RadioField,
    SelectField,
    SubmitButton,
    TextAreaField,
};
