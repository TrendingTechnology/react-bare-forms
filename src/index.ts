import {Form, FormConsumer} from "./form";
import {
    CheckBoxField,
    EmailField, FileField,
    PasswordField, RadioField, RadioGroup,
    SelectField,
    SubmitButton,
    TextAreaField,
    TextInputField
} from "./elements";
import {areFieldsEqual, isChecked, isEmailValid, isFieldEmpty, isFile, isRadioChecked} from "./validators";
import {createFileRef, getFileFromRef} from "./uncrontrolled";




export {
    // Uncontrolled
    createFileRef,
    getFileFromRef,
    // Validators
    areFieldsEqual,
    isChecked,
    isEmailValid,
    isFieldEmpty,
    isFile,
    isRadioChecked,
    // Form Components
    FormConsumer,
    Form,
    SubmitButton,
    TextInputField,
    EmailField,
    PasswordField,
    TextAreaField,
    CheckBoxField,
    SelectField,
    FileField,
    RadioField,
    RadioGroup,
}
