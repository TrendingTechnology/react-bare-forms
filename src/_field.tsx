import {FormContext, IFormContext} from "./form";
import {FormElementValidators, mergeDefaultCssWithProps} from "./_helpers";
import {ChangeEvent, default as React, ReactElement, useContext, useEffect} from "react";
import {
    IFileField,
    RadioGroupContext,
    TypeSelectCssSizeName
} from "./elements";
import {shouldUpdateRadioGroupContext} from "./_context_updaters";

interface IFieldClass<T> {
    create: (context: IFormContext) => ReactElement<T>;
    type: string;
    props: T;
}

/** @internal */
function _genericFormGroup<T extends any>(props: T, children: any) {
    return (
        <div className="form-group">
            {props.labelText && <label>{props.labelText}</label>}
            {children}
            {props.hint && <small className="form-text text-muted">{props.hint}</small>}
        </div>
    );
}

/** @internal */
abstract class _field<PropsType extends any> {

    type: string;

    props: PropsType;

    protected constructor(props: PropsType, type?: string) {
        this.type = type;
        this.props = props;
    }

    public createField(fieldCallback: Function) {
        const context: IFormContext = useContext(FormContext);
        const _validate = this.props.validators ?
            <FormElementValidators
                validators={this.props.validators}
                name={this.props.name}
                value={this.props.value}
                type="inputs"
            /> :
            null;
        if(context.bare) {
            return (<>{fieldCallback(context)}{_validate}</>);
        } else {
            return(<>{this.formGroup(fieldCallback(context))}{ _validate}</>);
        }
    }

    public abstract formGroup(children: any): ReactElement;

    public abstract getField(): (context: IFormContext) => ReactElement;

    static mergeDefaultCssWithProps(defaultValue: string, cssProps: any, bare: boolean): string {
        let cssStr = "";
        if(!bare) {
            cssStr += `${defaultValue} `;
        }
        if (cssProps) {
            cssStr += `${cssProps}`;
        }
        return cssStr;
    }

    public overrideEvent(e: any, value: any) {
        return {
            ...e,
            target: {
                ...e.target,
                value: !value,
            }
        }
    }

}

/** @internal */
export class InputField<T extends any> extends _field<T> implements IFieldClass<T> {

    constructor(type: string, props: T) {
        super(props, type);
        this.type = type;
        this.props = props;
    }

    public create() {
        return this.createField(this.getField());
    }

    public formGroup(children: any): ReactElement {
        return _genericFormGroup<T>(this.props, children);
    }

    public getField() {
        return (context: IFormContext) => {
            return <input
                type={this.type}
                value={context.state[this.props.name] || ""}
                onChange={(e) => context.updateParentState(e, this.props.name)}
                name={this.props.name}
                className={_field.mergeDefaultCssWithProps("form-control", this.props.className, context.bare)}
            />;
        }
    }

}

/** @internal */
export class CheckBoxField<T extends any> extends _field<T> implements IFieldClass<T> {

    constructor(type: string, props: T) {
        super(props, type);
        this.type = type;
        this.props = props;
    }

    public create() {
        return this.createField(this.getField());
    }

    public formGroup(children: any): ReactElement {
        return (
            <div className="form-group form-check">
                {children}
                {this.props.labelText && <label className="form-check-label">{this.props.labelText}</label>}
                {this.props.hint && <small className="form-text text-muted">{this.props.hint}</small>}
            </div>
        );
    }

    public getField() {
        return (context: IFormContext) => {
            return <input
                type={this.type}
                checked={context.state[this.props.name] || false}
                onChange={(e) => context.updateParentState(this.overrideEvent(e, context.state[this.props.name]), this.props.name)}
                name={this.props.name}
                className={_field.mergeDefaultCssWithProps("form-check-input", this.props.className, context.bare)}
            />;
        }
    }
}

/** @internal */
export class TextAreaField<T extends any> extends _field<T> implements IFieldClass<T> {

    constructor(props: T) {
        super(props);
        this.props = props;
    }

    public create() {
        return this.createField(this.getField());
    }

    public formGroup(children: any): ReactElement {
        return _genericFormGroup<T>(this.props, children);
    }

    public getField() {
        const {rows = 5} = this.props;
        return (context: IFormContext) => {
            return (
                <textarea
                    className={mergeDefaultCssWithProps("form-control", this.props.className, context.bare)}
                    rows={rows}
                    value={context.state[this.props.name] || ""}
                    onChange={(e) => context.updateParentState(e, this.props.name)}
                    name={this.props.name}
                />
            );
        }
    }
}

/** @internal */
export class RadioField<T extends any> extends _field<T> implements IFieldClass<T> {

    constructor(type: string, props: T) {
        super(props, type);
        this.type = type;
        this.props = props;
    }

    public create() {
        return this.createField(this.getField());
    }

    public formGroup(children: any): ReactElement {
        return (
            <div className="form-group form-check">
                {children}
                {this.props.labelText && <label className="form-check-label">{this.props.labelText}</label>}
                {this.props.hint && <small className="form-text text-muted">{this.props.hint}</small>}
            </div>
        );
    }

    public getField() {
        const radioContext: any = useContext(RadioGroupContext); // TODO - any
        return (context: IFormContext) => {
        const updateContexts = (e: React.ChangeEvent<any>, context: IFormContext) => {
            context.updateRadioGroupStateFromPassedInContext(
                this.overrideEvent(e, context.state[this.props.name]),
                this.props.name,
                radioGroup,
            );
            if(shouldUpdateRadioGroupContext(radioContext.children, context, radioContext.parent.name)) {
                useEffect(() => {
                    context.updateRadioGroupMetadata(radioContext.parent.name, radioContext.children);
                }, [radioContext]);

            }
        };
        const radioGroup = context.metadata.fieldGroups[radioContext.parent.name];
        return <input
            type={this.type}
            checked={context.state[this.props.name] || false}
            name={this.props.name}
            className={_field.mergeDefaultCssWithProps("form-check-input", this.props.className, context.bare)}
        />

        }
    }

}

/** @internal */
export class SelectField<T extends any> extends _field<T> implements IFieldClass<T> {

    constructor(props: T) {
        super(props);
        this.props = props;
    }

    public create() {
        return this.createField(this.getField());
    }

    public formGroup(children: any): ReactElement {
        return _genericFormGroup<T>(this.props, children);
    }

    public getField() {
        const {options = [], size = "default"} = this.props;
        return (context: IFormContext) => {

            return (
                <select
                    onChange={(e) => context.updateParentState(e, this.props.name)}
                    name={this.props.name}
                    className={_field.mergeDefaultCssWithProps(this.getSelectCssName(this.props.size), this.props.className, context.bare)}
                >
                    {options.map((optVal: string, i: number) => {
                        return <option value={optVal} key={i}>{optVal}</option>
                    })}
                </select>
            );
        }
    }

    private getSelectCssName(name: TypeSelectCssSizeName) {
        if(name === "default") {
            return "form-control"
        } else {
            return `form-control form-control-${name}`;
        }
    }
}

/** @internal */
export class FileField<T extends any> extends _field<T> implements IFieldClass<T> {

    constructor(props: T) {
        super(props);
        this.props = props;
    }

    public create() {
        return this.createField(this.getField());
    }

    public formGroup(children: any): ReactElement {
        return _genericFormGroup<T>(this.props, children);
    }

    public getField() {
        return (context: IFormContext) =>
            (<input
                ref={this.props.ref}
                type="file"
                className={_field.mergeDefaultCssWithProps("form-control-file", this.props.className, context.bare)}
            />);
    }
}
