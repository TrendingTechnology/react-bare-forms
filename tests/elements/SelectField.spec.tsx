import * as React from "react";
import {create} from "react-test-renderer";
import {Form, SelectField} from "../../src";


describe("#<SelectField>", () => {
    it("should render props", () => {
        let state = {
            select_data_id: undefined as any,
        };

        let component = create(
            <Form state={state}>
                <SelectField
                    size="lg"
                    value={state.select_data_id}
                    name="fruitChoice"
                    options={["banana", "apple", "orange"]}
                />
            </Form>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
    it("should render props from object keys", () => {
        let selectData: Array<{id: number, name: string}> = [
            {id: 1, name: "first"},
            {id: 2, name: "second"},
            {id: 3, name: "third"},
        ];
        let state = {
            select_data_id: undefined as any,
        };

        let component = create(
            <Form state={state}>
                <SelectField
                    size="lg"
                    value={state.select_data_id}
                    name="fruitChoice"
                    objectKey="id"
                    objectValue="name"
                    options={selectData}
                />
            </Form>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
