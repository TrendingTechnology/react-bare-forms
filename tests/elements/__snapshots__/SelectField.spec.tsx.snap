// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#<SelectField> should render props 1`] = `
<form
  onSubmit={[Function]}
  state={
    Object {
      "select_data_id": undefined,
    }
  }
>
  <div
    className="form-group"
  >
    <select
      className="form-control form-control-lg "
      name="fruitChoice"
      onChange={[Function]}
    >
      <option
        value="banana"
      >
        banana
      </option>
      <option
        value="apple"
      >
        apple
      </option>
      <option
        value="orange"
      >
        orange
      </option>
    </select>
  </div>
</form>
`;

exports[`#<SelectField> should render props from object keys 1`] = `
<form
  onSubmit={[Function]}
  state={
    Object {
      "select_data_id": undefined,
    }
  }
>
  <div
    className="form-group"
  >
    <select
      className="form-control form-control-lg "
      name="fruitChoice"
      onChange={[Function]}
    >
      <option
        value={1}
      >
        first
      </option>
      <option
        value={2}
      >
        second
      </option>
      <option
        value={3}
      >
        third
      </option>
    </select>
  </div>
</form>
`;
